src/vm/jitinterface.cpp

```diff
@@ -8760,7 +8773,17 @@ static CORINFO_METHOD_HANDLE resolveVirtualMethodHelper(MethodDesc* callerMethod
         // safely devirtualize.
         if (ownerType != nullptr)
         {
-            pDevirtMD = pDerivedMT->GetMethodDescForInterfaceMethod(GetTypeFromContext(ownerType), pBaseMD);
+            TypeHandle OwnerClsHnd = GetTypeFromContext(ownerType);
+            MethodTable* pOwnerMT = OwnerClsHnd.GetMethodTable();
+
+            // If the derived class is a shared class, make sure the
+            // owner class is too.
+            if (pDerivedMT->IsSharedByGenericInstantiations())
```

Comment:
Agree this is confusing. The jit isn't given or getting exact types in many cases. But when inlining a shared method into an unshared one, it ends up with the exact method type. So it can end up with an mixed view of types when devirtualizing a call in an inlinee.

For virtual calls this mixed view of types didn't matter since an exact type and its canonical type have the same vtable.

For invariant interface calls the mixed view likewise should not have mattered since the instance types do not figure in target resolution. I suspect one could add an equivalence check somewhere in the interface call resolution path that could fix the mixed case. However I was reluctant to make any changes there. Hence this fix.

By using shared types, we may lose the ability to devirtualize variant interface call since we can no longer reason about the relationships of the instance types. However it turns out we'd already lost those with the `CanCastToInterface` call above.

So the only diffs from this change are from no longer devirtualizing interface calls where the target is a final method; the changes on the VM side don't alter the jit's codegen.

Longer term I hope to selectively uplevel the types visible to the jit.